lib/engine-object-file/src/engine.rs

```diff
@@ -0,0 +1,252 @@
+
+/// The inner contents of `ObjectFileEngine`
+pub struct ObjectFileEngineInner {
+    /// The compiler
+    #[cfg(feature = "compiler")]
+    compiler: Option<Box<dyn Compiler + Send>>,
+    /// The WebAssembly features to use
+    #[cfg(feature = "compiler")]
+    features: Features,
+    /// Pointers to trampoline functions used to enter particular signatures
+    trampolines: HashMap<VMSharedSignatureIndex, VMTrampoline>,
+    /// The signature registry is used mainly to operate with trampolines
+    /// performantly.
+    signatures: SignatureRegistry,
+    /// The prefixer returns the a String to prefix each of
+    /// the functions in the shared object generated by the `ObjectFileEngine`,
+    /// so we can assure no collisions.
+    prefixer: Option<Box<dyn Fn(&[u8]) -> String + Send>>,
+}
+
+impl ObjectFileEngineInner {
+    /// Gets the compiler associated to this engine.
+    #[cfg(feature = "compiler")]
+    pub fn compiler(&self) -> Result<&dyn Compiler, CompileError> {
+        if self.compiler.is_none() {
+            return Err(CompileError::Codegen("The `ObjectFileEngine` is operating in headless mode, so it can only execute already compiled Modules.".to_string()));
+        }
+        Ok(&**self
+            .compiler
+            .as_ref()
+            .expect("Can't get compiler reference"))
+    }
+
+    #[cfg(feature = "compiler")]
+    pub(crate) fn get_prefix(&self, bytes: &[u8]) -> String {
+        if let Some(prefixer) = &self.prefixer {
+            prefixer(&bytes)
+        } else {
+            "".to_string()
+        }
+    }
+
+    #[cfg(feature = "compiler")]
+    pub(crate) fn features(&self) -> &Features {
+        &self.features
+    }
+
+    /// Validate the module
+    #[cfg(feature = "compiler")]
+    pub fn validate<'data>(&self, data: &'data [u8]) -> Result<(), CompileError> {
+        self.compiler()?.validate_module(self.features(), data)
+    }
+
+    /// Validate the module
+    #[cfg(not(feature = "compiler"))]
+    pub fn validate<'data>(&self, _data: &'data [u8]) -> Result<(), CompileError> {
+        Err(CompileError::Validate(
+            "The `ObjectFileEngine` is not compiled with compiler support, which is required for validating".to_string(),
+        ))
+    }
+
+    /// Shared signature registry.
+    pub fn signatures(&self) -> &SignatureRegistry {
+        &self.signatures
+    }
+
+    /// Gets the trampoline pre-registered for a particular signature
+    pub fn trampoline(&self, sig: VMSharedSignatureIndex) -> Option<VMTrampoline> {
+        self.trampolines.get(&sig).cloned()
+    }
+
+    pub(crate) fn add_trampoline(&mut self, func_type: &FunctionType, trampoline: VMTrampoline) {
+        let index = self.signatures.register(&func_type);
+        // We always use (for now) the latest trampoline compiled
+        // TODO: we need to deallocate trampolines as the compiled modules
+        // where they belong become unallocated.
+        self.trampolines.insert(index, trampoline);
```

Comment:
Still todo?

Reply:
Yeah looks like it... this code is duplicated from `engine-native` and seems to be just an oversight on both.  I don't really have context on it right now